Add dts yaml support for same sensor on i2c or spi

galak opened the pull request:
Update how we handle the YAML files in the extract script to support the same compatible for a sensor being on either I2C or SPI.  We still have unique YAML files, but the DTS compatible can now be the same.

galak committed:
scripts: extract_dts_includes: Handle YAML inheritance when we load

Move the handling of 'inherits' in the YAML to right when we load.  This
allows us to remove yaml_collapse function and will allow us to look
into the yaml right after we load it for things like if the yaml is for
a child bus.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: extract_dts_includes: remove passing yaml param to functions

Various functions like add_prop_aliases, extract_single_prop,
extract_single, reg.extract, compatible.extract, default.extract, and
flash.extract don't use the yaml argument that is passed to them.
Remove passing the argument, this is the first step in making access to
the yaml node information go through an accessor function.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: extract_dts_includes: Add accessor function to access binding

Add accessor functions get_binding and get_binding_compats to
encapsulate access to binding database info.  Cleanup all direct access
to binding info and replace with calls to get_binding() and
get_binding_compats()

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
sensor: lsm6dsl: Update DTS support

Now that we can support the same compatible but different bus types,
update the LSM6DSL support to utilize the same compatible for either I2C
or SPI.  We rename the i2c binding file to st,lsm6dsl-i2c.yaml just to
be a bit more clear.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -60,7 +60,9 @@ def bindings(cls, compatibles, yaml_dirs):
-        yaml_list = {}
+        yaml_list['node'] = {}
+        yaml_list['bus'] = {}
+        yaml_list['compat'] = []
```
Hmm... this confuses me though. Doesn't `yaml_list` just map compats (from lines like `constraint: "nxp,imx-lpspi"`) to the parsed yaml from that file?

Would it make sense to have these as separate variables, instead of as keys in `yaml_list`/`compat_to_yaml`?

b0661 replied:
Renaming yaml_list to `bindings` would make the intention a little bit clearer.

galak replied:
agreed

ulfalizer replied:
Fine with me

ulfalizer commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -73,7 +75,14 @@ def bindings(cls, compatibles, yaml_dirs):
-                                yaml_list[c] = l
+                                if c not in yaml_list['compat']:
+                                    yaml_list['compat'].append(c)
+                                if ('parent' in l):
```
Redundant parentheses

galak replied:
Fixed

ulfalizer reviewed (commented):
Just some random comments. Not very familiar with this code.

b0661 commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -744,9 +737,9 @@ def main():
     get_chosen(dts['/'])
 
-    yaml_list = load_yaml_descriptions(dts, args.yaml)
+    extract.globals.yaml_list = load_yaml_descriptions(dts, args.yaml)
```
This makes yaml_list a global with a hardly understandable variable name. What about using `bindings` instead?

galak replied:
Agreed.

b0661 reviewed (commented):
I would prefer to stop using the term `yaml` and start replacing it by `bindings` where appropriate.

erwango commented on boards/arm/96b_argonkey/96b_argonkey.dts:
```diff
@@ -60,7 +60,7 @@
-		compatible = "st,lsm6dsl-spi";
+		compatible = "st,lsm6dsl";
 		reg = <1>;
 		spi-max-frequency = <1000000>;
```
IIUC, this whole change allows to have these bus related properties documented by the binding depending on the bus. Is that compatible with device tree usage in eg Linux?

I'm still not convinced how this would work at driver level. To me we should be able to build matching sensor driver using "compatible".
When "compatible" carried SPI/I2C inforamtion, we could have used #idef DT_COMPAT_ST_LSM6DL_SPI / DT_COMPAT_ST_LSM6DL_I2C to know if driver should be compiled with SPI or I2C fields.
Now this information is no more present we'll need to have a new piece of info to know how to build the driver with fields matching the bus.

erwango replied:
but then it should be available as a compilation input, as a configuration input.

pabigot replied:
The bus type can be inferred by looking at the parent node, and the tooling generate the appropriate flags.  The BME280 supports SPI and I2C, and the Linux binding documentation doesn't represent which is being used: it just provides two drivers, each of which are compatible with `bosch,bme280`.

The inference would be simplified by marking all I2C bus nodes as compatible with `zephyr,i2c` and similarly for other generic Zephyr drivers.  This is something I'd like done regardless.

galak replied:
> IIUC, this whole change allows to have these bus related properties documented by the binding depending on the bus. Is that compatible with device tree usage in eg Linux?

Yes, if you look at how various sensor drivers in linux you'll see they use the same compatible regardless of what bus the device is on.

galak replied:
> The inference would be simplified by marking all I2C bus nodes as compatible with `zephyr,i2c` and similarly for other generic Zephyr drivers. This is something I'd like done regardless.

I don't see us ever doing this, as we should be aiming for the device tree's to have less zephyr specific details not more.  The intent would be to know that a bus is I2C or SPI via the binding YAML info.

avisconti reviewed (commented):
Seems fine, but I will have to review some of the changes I have done (am doing) on sensor part.
So, basically there will still be two yaml file (say lsm6dsl-spi.yaml and lsm6dsl-i2c.yaml) using
same compatible string inside.

erwango reviewed (commented):
Clean up is good but I'm still not convinced with the main idea of the change.
At the very least it misses a way to instruct how to build the sensor driver with the ad-hoc structs, since the info is no more carried by "compatible".

galak commented on scripts/dts/extract/globals.py:
```diff
@@ -286,3 +287,24 @@ def add_prop_aliases(node_address, yaml,
         if (old_alias_names and old_alias_label != prop_label):
             prop_aliases[old_alias_label] = prop_label
+
+def get_yaml_node(node_address):
```
Changing to get_binding()

galak commented on scripts/dts/extract/globals.py:
```diff
@@ -286,3 +287,24 @@ def add_prop_aliases(node_address, yaml,
+
+    return yaml_node
+
+def get_yaml_compats():
```
changing to get_binding_compats

galak committed:
scripts: extract_dts_includes: Pass node address to get_binding

Change get_binding to take the node address instead of the
compatiable.  This is in prep for having get_binding be able to look
at the parent of the node and determine which bus specific binding to
utilize (for cases like sensors that support both I2C & SPI).

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on scripts/dts/extract/globals.py:
```diff
@@ -286,3 +287,9 @@ def add_prop_aliases(node_address,
+    return yaml_list[compat]
+
+def get_yaml_compats():
+    return yaml_list.keys()
```
Going with binding based on @b0661 comments.

galak commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -664,7 +673,7 @@ def load_yaml_descriptions(dts, yaml_dirs):
 
     yaml_list = Bindings.bindings(compatibles, yaml_dirs)
-    if yaml_list == {}:
+    if yaml_list['node'] == {}:
```
chagned up

galak committed:
scripts: extract_dts_includes: Split out yaml_list info

Have the yaml_list now have 'node', 'bus', and 'compat' information
instead of just the 'node' info.  For now we have all YAML nodes under
yaml_list['node'], however this will change to support sensors that
can be on I2C or SPI, since they will have the same compat, but
different YAML nodes.

This also introduces yaml_list['bus'] which has YAML nodes for any
compatiables on a given 'parent' bus (ie, 'i2c', or 'spi').

Finally, we also have yaml_list['compat'] with all the compatiables
that we parsed and matched to a YAML.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: extract_dts_includes: Support same sensor on different bus

There are a number of cases in which we a sensor can be either connected
on I2C or SPI.  We've been treating these cases as different compatiable
properties, but they really should use the same compatiable and just
determine the info based on the parent bus in the device tree.  We can
now support having two different binding files for the same compatiable
to handle the case of a sensor supporting either I2C or SPI as how its
connected.

We put "sensor" nodes in a bus specific dict bus_bindings[bus] that
we can than lookup later based on the DTS and determining the bus type
that the "sensor" node is on.

Fixes #11375

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
Pushed an update to fix up comments from @ulfalizer and @b0661

pabigot commented on scripts/dts/extract/clocks.py:
```diff
@@ -42,8 +41,8 @@ def _extract_consumer(self, node_address, yaml, clocks, def_label):
-                clock_provider_compat = get_compat(clock_provider_node_address)
-                clock_provider_bindings = yaml[clock_provider_compat]
+                clock_provider_bindings = get_binding(
+                                            clock_provider_node_address)
```
Off topic but this sort of thing strengthens my belief that a strict 80-character limit on line length is decreasing code clarity more than increasing it.  Especially when EOL escapes used to meet the requirement aren't consistent (see next line).

carlescufi replied:
The 80 char limit is not enforced. `checkpatch` will emit a warning, but CI will not fail. Contributors are free to exceed the limit when it makes sense to do so.

pabigot reviewed (commented):
I still haven't grasped the design goals/constraints of all this tooling, but within those limits I agree with the purpose of this commit and have no concerns about the changes.

galak committed:
scripts: extract_dts_includes: Generate device bus info define

For devices on buses like I2C or SPI its possible that a sensor might
support being on either bus type.  In the dts we can tell this by
looking at the parent bus of the device.  Its useful for a driver to
know this so can we build support into the driver accordingly.

For example if the LSM6DSL sensor ("st,lsm6dsl") is in the dts on a spi
bus we now generate:

	#define DT_ST_LPS22HB_PRESS_BUS_SPI	1

Its possible that a system exists in which multiple of the same sensor
exist but on different busses, so drivers should handle that case
accordingly.  For the LSM6DSL example we'd end up with:

	#define DT_ST_LPS22HB_PRESS_BUS_I2C	1
	#define DT_ST_LPS22HB_PRESS_BUS_SPI	1

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
